src/bytes_mut.rs

```diff
@@ -1266,6 +1288,38 @@ fn original_capacity_from_repr(repr: usize) -> usize {
     1 << (repr + (MIN_ORIGINAL_CAPACITY_WIDTH - 1))
 }
 
+#[inline]
+fn check_capacity_overflow(capacity: usize) {
+    if mem::size_of::<usize>() < 8 && capacity > isize::MAX as usize {
+        capacity_overflow()
+    }
+}
+
+fn capacity_overflow() -> ! {
+    panic!("capacity overflow")
+}
+
+#[inline]
+fn alloc_buf(capacity: usize) -> NonNull<u8> {
+    if capacity == 0 {
+        NonNull::dangling()
+    } else {
+        check_capacity_overflow(capacity);
+        unsafe {
+            let layout = Layout::from_size_align_unchecked(capacity, 1);
+            non_null(alloc(layout), layout)
+        }
+    }
+}
+
+#[inline]
+fn non_null(ptr: *mut u8, layout: Layout) -> NonNull<u8> {
```

Comment:
It might be worth having a comment here (and on the already-present `vptr` function?) explaining when they're supposed to be used. It looks like `vptr` is called to construct a `NonNull` from a raw ptr which is assumed to be an already valid pointer to a `Vec`, while this function is called to construct a `NonNull` from a raw ptr which was returned by the allocator? But, I am not sure if this distinction is clear on the first read-through.

Reply:
Is this error, https://github.com/rust-lang/rust/blob/8ad7bc3f428300aee6764f6e23527e19eb235e81/src/liballoc/alloc.rs#L177.
In `vptr`, this error has already been checked since it comes from a `Vec`.